### PR TITLE
Add 'output_json_as_list' option to CLI (WIP)

### PR DESCRIPTION
**Pull request by martynassateika**
Everything about this is still WIP, opening it up early in case I'm misunderstanding something.

By the way I needed to change this line on Windows to

```python
opened = open(filename, "w")
```

to get the tests to pass:

https://github.com/stevearc/dql/blob/6b0ab8df07afd226816937ee98b4ef73187206fc/dql/engine.py#L581

Fixes https://github.com/stevearc/dql/issues/21.

### CODE REVIEW

File: tests/test_queries.py [hunk 1]
```diff
@@ -495,6 +495,59 @@ def test_select_non_projected(self):
         self.assertEqual(list(ret), [{"bar": 2}])
 
 
+class TestSaveJson(BaseSystemTest):
+
+    """ Tests that involve saving the results into a JSON file """
+
+    def setUp(self):
+        super(TestSaveJson, self).setUp()
+        self.test_file_name = 'out.json'
```
Comment (martynassateika) [inline tests/test_queries.py]: We want to write to a temporary file ideally, however if the parser receives a full Windows path it blows up :-(

File: tests/test_queries.py [hunk 2]
```diff
@@ -495,6 +495,59 @@ def test_select_non_projected(self):
...
+        self.make_table()
+        self.query("INSERT INTO foobar (id, bar) VALUES ('a', 1), ('b', 2)")
+        self.query("SCAN * FROM foobar SAVE " + self.test_file_name)
+        contents = self._get_file_contents()
+        self.assertEqual(
+            contents,
+            '[{"bar":2,"id":"b"},\n{"bar":1,"id":"a"}]\n'
+        )
+
+    def test_three_results(self):
```
Comment (martynassateika) [inline tests/test_queries.py]: I went up to three to ensure there's a case where the opening and closing bracket are separted by at least one line in the middle.

File: tests/test_queries.py [hunk 3]
```diff
@@ -495,6 +495,59 @@ def test_select_non_projected(self):
...
+        """ JSON output with three results """
+        self.make_table()
+        self.query(
+            "INSERT INTO foobar (id, bar) VALUES ('a', 1), ('b', 2), ('c', 3)"
+        )
+        self.query("SCAN * FROM foobar SAVE " + self.test_file_name)
+        contents = self._get_file_contents()
+        self.assertEqual(
+            contents,
+            '[{"bar":2,"id":"b"},\n{"bar":3,"id":"c"},\n{"bar":1,"id":"a"}]\n'
```
Comment (martynassateika) [inline tests/test_queries.py]: Hmm, do I need an ORDER BY here or is the order of results always going to be consistent.

File: dql/cli.py [hunk 4]
```diff
@@ -396,6 +397,16 @@ def complete_opt_allow_select_scan(self, text, *_):
         """ Autocomplete for allow_select_scan option """
         return [t for t in ("true", "false", "yes", "no") if t.startswith(text.lower())]
 
+    def opt_output_json_as_list(self, output):
+        """ Set option output_json_as_list """
+        output = output.lower() in ("true", "t", "yes", "y")
+        self.conf["output_json_as_list"] = output
+        self.engine.output_json_as_list = output
```
Comment (stevearc) [inline dql/cli.py]: You may want to set the value on the engine during startup as well, with the value loaded from the config. Probably around here: https://github.com/stevearc/dql/blob/6b0ab8df07afd226816937ee98b4ef73187206fc/dql/cli.py#L202

File: tests/test_queries.py (hunk 1 above)
Comment (stevearc) [reply tests/test_queries.py]: I think that all you have to do to support a windows file path is to add `\` and `:` to the filename regex here:
https://github.com/stevearc/dql/blob/6b0ab8df07afd226816937ee98b4ef73187206fc/dql/grammar/common.py#L168

if you want to try that

File: tests/test_queries.py (hunk 3 above)
Comment (stevearc) [reply tests/test_queries.py]: It's defined by dynamodb, so in practice it's pretty consistent, but it can change at any time

**Comment by martynassateika** [comment]
Hi Steven. I've been doing some cleanup of my open PRs, and realized this was open for nearly 5 years now. I haven't used DQL in years, so it's probably better to just close this than dig into the code again 🙏

**Pull request closed without merging**